client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Selects nodes that can receive requests. Used to keep requests away
+ * from master nodes or to send them to nodes with a particular attribute.
+ * Use with {@link Request#setNodeSelector(NodeSelector)}.
+ */
+public interface NodeSelector {
+    /**
+     * Select the {@link Node}s to which to send requests. This is called with
+     * a list of {@linkplain Node}s in the order that the rest client would
+     * prefer to use them and it should remove nodes from the list that should
+     * not receive the request.
+     * <p>
+     * This may be called twice per request: first for "living" nodes that
+     * have not been blacklisted by previous errors. In this case the order
+     * of the nodes is the order in which the client thinks that they should
+     * be tried next. If the selector removes all nodes from the list or if
+     * there aren't any living nodes then the the client will call this method
+     * with a list of "dead" nodes. In this case the list is sorted "soonest
+     * to be revived" first. In this case the rest client will only attempt
+     * the first node.
+     * <p>
+     * Implementations <strong>may</strong> reorder the list but they should
+     * be careful in doing so as the original order is important (see above).
+     * An implementation that sorts list consistently will consistently send
+     * requests to s single node, overloading it. So implementations that
+     * reorder the list should take the original order into account
+     * <strong>somehow</strong>.
```

Comment:
You argued strongly a few months ago to allow it, mostly on the grounds that something like rack awareness features would want to sort them like `<on_the_rack>, <off_the_rack>`.

Reply:
oops. that's what happens. I forget things. I wouldn't say I argued strongly about ordering necessarily, I just wanted the selector to get a list of all the node candidates rather than a single node. That way it can decide what to do and act as a preference, meaning if there's not nodes in the rack, it's ok to go to an alive node in another rack rather than trying to revive a node. That can be achieved without allowing the selector to mess with ordering?

Reply:
What about something like this:
```
diff --git a/client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java b/client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java
index e9fbc9b6999..273cecac7ef 100644
--- a/client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java
+++ b/client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java
@@ -54,15 +54,26 @@ public interface NodeSelector {
      *      selector approves of, in the order that the selector would prefer
      *      to use them.
      */
-    List<Node> select(List<Node> nodes);
+    Selector prepare(List<Node> nodes);
+
+    interface Selector {
+        boolean select(Node node);
+    }
 
     /**
      * Selector that matches any node.
      */
     NodeSelector ANY = new NodeSelector() {
+        private final Selector SELECTOR = new Selector() {
+            @Override
+            public boolean select(Node node) {
+                return true;
+            }
+        };
+
         @Override
-        public List<Node> select(List<Node> nodes) {
-            return nodes;
+        public Selector select(List<Node> nodes) {
+            return SELECTOR;
         }
 
         @Override
@@ -77,16 +88,19 @@ public interface NodeSelector {
      * role. It does not reorder the nodes sent to it.
      */
     NodeSelector NOT_MASTER_ONLY = new NodeSelector() {
-        @Override
-        public List<Node> select(List<Node> nodes) {
-            List<Node> subset = new ArrayList<>(nodes.size());
-            for (Node node : nodes) {
-                if (node.getRoles() == null) continue;
-                if (false == node.getRoles().isMasterEligible() || node.getRoles().isData()) {
-                    subset.add(node);
+        private final Selector SELECTOR = new Selector() {
+            @Override
+            public boolean select(Node node) {
+                if (node.getRoles() == null) {
+                    return false;
                 }
+                return false == node.getRoles().isMasterEligible() || node.getRoles().isData();
             }
-            return subset;
+        };
+
+        @Override
+        public Selector select(List<Node> nodes) {
+            return SELECTOR;
         }

         @Override
diff --git a/client/rest/src/main/java/org/elasticsearch/client/RestClient.java b/client/rest/src/main/java/org/elasticsearch/client/RestClient.java
index 1e6a477e215..d3c613c5535 100644
--- a/client/rest/src/main/java/org/elasticsearch/client/RestClient.java
+++ b/client/rest/src/main/java/org/elasticsearch/client/RestClient.java
@@ -648,9 +648,14 @@ public class RestClient implements Closeable {
              */
             // TODO this is going to send more requests to nodes right *after* a node that the selector removes
             Collections.rotate(livingNodes, lastNodeIndex.getAndIncrement());
-            List<Node> selectedLivingNodes = nodeSelector.select(livingNodes);
-            if (false == selectedLivingNodes.isEmpty()) {
-                return selectedLivingNodes;
+            Selector selector = nodeSelector.selector(livingNodes);
+            for (Iterator<Node> itr = livingNodes.iterator(); itr.hasNet();) {
+                if (false == selector.select(itr.next())) {
+                    itr.remove();
+                }
+            }
+            if (false == livingNodes.isEmpty()) {
+                return livingNodes;
             }
         }
 
@@ -677,7 +682,12 @@ public class RestClient implements Closeable {
             for (DeadNodeAndRevival n : deadNodes) {
                 selectedDeadNodes.add(n.node);
             }
-            selectedDeadNodes = nodeSelector.select(selectedDeadNodes);
+            Selector selector = nodeSelector.selector(selectedDeadNodes);
+            for (Iterator<Node> itr = selectedDeadNodes.iterator(); itr.hasNet();) {
+                if (false == selector.select(itr.next())) {
+                    itr.remove();
+                }
+            }
```

The NodeSelector can still look at all the nodes before it makes a decision but it can't reorder.

Reply:
I do not love that two interfaces need to be implemented, especially the `select` method that returns a `Selector` feels a bit of a weird API to me. What do you think of accepting and returning a `Collection` or a `Set` instead?

Reply:
I think a `Set` is pretty heavy. What if I send it an `Iterable`? They can iterate it once to figure out everything in the list and then they can iterate it again to remove things. And we already **have** an Iterable on the call site so I don't need to build any more objects.

Reply:
++